src/System.Windows.Forms/src/System/Windows/Forms/OLE/DataObject.cs

```diff
@@ -100,13 +102,148 @@ internal IDataObject TryUnwrapInnerIDataObject()
+    ///   and you are using a custom format or <see cref="DataFormats.Serializable"/>
+    ///   it is recommended to use the <see cref="SetData(string, object?)"/> APIs to avoid unnecessary overhead.
+    ///  </para>
+    ///  <para>
+    ///   The default behavior of <see cref="JsonSerializer"/> is used to serialize the data.
+    ///  </para>
+    ///  <para>
+    ///   See
+    ///   <see href="https://learn.microsoft.com/dotnet/standard/serialization/system-text-json/how-to#serialization-behavior"/>
+    ///   and <see href="https://learn.microsoft.com/dotnet/standard/serialization/system-text-json/reflection-vs-source-generation#metadata-collection"/>
+    ///   for more details on default <see cref="JsonSerializer"/> behavior.
+    ///  </para>
+    ///  <para>
+    ///   If custom JSON serialization behavior is needed, manually JSON serialize the data and then use <see cref="SetData(object?)"/>,
+    ///   or create a custom <see cref="Text.Json.Serialization.JsonConverter"/>, attach the
+    ///   <see cref="Text.Json.Serialization.JsonConverterAttribute"/>, and then recall this method.
+    ///   See <see href="https://learn.microsoft.com/dotnet/standard/serialization/system-text-json/converters-how-to"/> for more details
+    ///   on custom converters for JSON serialization.
+    ///  </para>
+    /// </remarks>
+    [RequiresUnreferencedCode("Uses default System.Text.Json behavior which is not trim-compatible.")]
+    public void SetDataAsJson<T>(string format, bool autoConvert, T data)
+    {
+        if (typeof(T) == typeof(DataObject))
+        {
+            // loni TODO: localize string.
+            throw new InvalidOperationException($"DataObject will serialize as empty. JSON serialize the data within {nameof(data)}, then use {nameof(SetData)} instead.");
+        }
+
+        SetData(format, autoConvert, TryJsonSerialize(format, data));
+    }
+
+    /// <summary>
+    ///  JSON serialize the data only if the format is not a restricted deserialization format and the data is not an intrinsic type.
+    /// </summary>
+    /// <returns>
+    ///  The passed in <paramref name="data"/> as is if the format is restricted. Otherwise the JSON serialized <paramref name="data"/>.
+    /// </returns>
+    private static object TryJsonSerialize<T>(string format, T data)
+    {
+        format.OrThrowIfNull(nameof(format));
+        data.OrThrowIfNull(nameof(data));
+        return IsRestrictedFormat(format) || Composition.Binder.IsKnownType<T>()
+            ? data
+            : new JsonData<T>() { JsonBytes = JsonSerializer.SerializeToUtf8Bytes(data) };
+    }
+
+    /// <summary>
+    ///  Check if the <paramref name="format"/> is one of the restricted formats, which formats that
+    ///  correspond to primitives or are pre-defined in the OS such as strings, bitmaps, and OLE types.
+    /// </summary>
+    internal static bool IsRestrictedFormat(string format) => RestrictDeserializationToSafeTypes(format)
```

Comment:
```suggestion
    private static bool IsRestrictedFormat(string format) => RestrictDeserializationToSafeTypes(format)
```
This is used in the DataObject class and its inner classes only